File: packages/protocol/src/pairing-coordinator.ts
/**
 * Pairing ceremony coordinator for TypeScript / Web (V19-PR08).
 * Drives the multi-step pairing ceremony on both initiator and responder sides,
 * enforcing mutual key confirmation, tombstone checks, and transactional persistence.
 *
 * Symmetrical counterpart to Go `packages/engine/trust/pairing_coordinator.go`.
 */

import { bytesToHex, constantTimeEqual, hexToBytes } from './bytes.js';
import { ERR_KEY_CONFLICT, ERR_LABEL_CONFLICT, ERR_TRUSTED_PEER_REVOKED } from './errors.js';
import type { DeviceIdentity } from './identity.js';
import {
  MSG_PAIRING_CONFIRM,
  MSG_PAIRING_REQUEST,
  MSG_PAIRING_RESPONSE,
  createPairingConfirm,
  createPairingRequest,
  createPairingResponse,
  derivePairCredential,
  verifyPairingConfirm,
  verifyPairingRequest,
  verifyPairingResponse,
  type PairingConfirm,
  type PairingMessage,
} from './pairing.js';
import type { TombstoneStore } from './tombstone-store.js';
import {
  RELATIONSHIP_CONTACT,
  type TrustPolicy,
  type TrustRecord,
  type TrustStore,
} from './trust-store.js';

export interface PairingSessionConfig {
  deviceName: string;
  capabilities: string[];
  masterKey: Uint8Array;
  autoAccept?: boolean;
  destDir?: string;
}

export interface PairingResult {
  peerRecord: TrustRecord;
  kPair: Uint8Array;
  credRef: string;
}

export interface SecretStore {
  setPairSecret(deviceId: string, secret: Uint8Array): Promise<void>;
  deletePairSecret(deviceId: string): Promise<void>;
}

export interface PairingTransport {
  sendMessage(msg: PairingMessage): Promise<void>;
  receiveMessage(): Promise<PairingMessage | string | Uint8Array>;
}

export function parsePairingMessage(raw: PairingMessage | string | Uint8Array): PairingMessage {
  if (typeof raw === 'string') {
    return JSON.parse(raw) as PairingMessage;
  }
  if (raw instanceof Uint8Array) {
    const text = new TextDecoder().decode(raw);
    return JSON.parse(text) as PairingMessage;
  }
  return raw;
}

export class PairingCoordinator {
  private readonly trustStore: TrustStore;
  private readonly secretStore?: SecretStore | undefined;
  private readonly tombstoneStore?: TombstoneStore | undefined;

  constructor(trustStore: TrustStore, secretStore?: SecretStore, tombstoneStore?: TombstoneStore) {
    this.trustStore = trustStore;
    this.secretStore = secretStore;
    this.tombstoneStore = tombstoneStore;
  }

  async initiatePairing(
    transport: PairingTransport,
    cfg: PairingSessionConfig,
    id: DeviceIdentity,
  ): Promise<PairingResult> {
    if (!transport) throw new Error('pairing transport required');
    if (!cfg.masterKey || cfg.masterKey.length === 0) throw new Error('master key required');
    if (!id) throw new Error('local identity required');

    // 1. Create and send PairingRequest
    const req = await createPairingRequest(id, cfg.deviceName, cfg.capabilities, cfg.masterKey);
    await transport.sendMessage(req);

    // 2. Receive and verify PairingResponse
    const rawResp = await transport.receiveMessage();
    const respMsg = parsePairingMessage(rawResp);
    if (!respMsg || respMsg.type !== MSG_PAIRING_RESPONSE) {
      throw new Error('expected pairing response message');
    }

    const reqNonce = hexToBytes(req.nonce);
    const { publicKey: peerPub, nonce: respNonce } = await verifyPairingResponse(
      respMsg,
      reqNonce,
      cfg.masterKey,
    );

    // 3. Derive pairwise credential
    const { kPair, credRef } = await derivePairCredential(
      cfg.masterKey,
      reqNonce,
      respNonce,
      id.publicKey,
      peerPub,
    );

    // 4. Verify no key or label conflict in trust store
    try {
      await this.checkConflict(respMsg.device_id, respMsg.device_name, peerPub);
    } catch (err) {
      await this.sendRejection(transport);
      throw err;
    }

    // 5. Send and receive PairingConfirm
    const confirm = await createPairingConfirm(kPair, respMsg.device_id, true);
    await transport.sendMessage(confirm);

    const rawPeerConf = await transport.receiveMessage();
    const peerConfMsg = parsePairingMessage(rawPeerConf);
    if (!peerConfMsg || peerConfMsg.type !== MSG_PAIRING_CONFIRM) {
      throw new Error('expected pairing confirm message');
    }
    await verifyPairingConfirm(peerConfMsg, kPair, id.deviceId);

    // 6. Record peer in local trust store & secret store (with rollback)
    const now = new Date().toISOString();
    const policy: TrustPolicy = {
      autoAccept: !!cfg.autoAccept,
      ...(cfg.autoAccept && cfg.destDir ? { autoAcceptDestDir: cfg.destDir } : {}),
      maxFileSizeBytes: 10 * 1024 * 1024 * 1024,
    };

    const record: TrustRecord = {
      deviceId: respMsg.device_id,
      publicKey: bytesToHex(peerPub),
      localLabel: respMsg.device_name,
      pairCredentialRef: credRef,
      relationship: RELATIONSHIP_CONTACT,
      capabilities: respMsg.capabilities || [],
      firstSeenAt: now,
      lastSeenAt: now,
      revoked: false,
      policy,
    };

    if (this.secretStore) {
      await this.secretStore.setPairSecret(record.deviceId, kPair);
    }

    try {
      await this.trustStore.addOrUpdateDevice(record);
    } catch (err) {
      if (this.secretStore) {
        try {
          await this.secretStore.deletePairSecret(record.deviceId);
        } catch {
          // ignore rollback failure
        }
      }
      throw new Error(`save trusted device: ${(err as Error).message}`);
    }

    return {
      peerRecord: record,
      kPair,
      credRef,
    };
  }

  async acceptPairing(
    transport: PairingTransport,
    cfg: PairingSessionConfig,
    id: DeviceIdentity,
  ): Promise<PairingResult> {
    if (!transport) throw new Error('pairing transport required');
    if (!cfg.masterKey || cfg.masterKey.length === 0) throw new Error('master key required');
    if (!id) throw new Error('local identity required');

    // 1. Receive and verify PairingRequest
    const rawReq = await transport.receiveMessage();
    const reqMsg = parsePairingMessage(rawReq);
    if (!reqMsg || reqMsg.type !== MSG_PAIRING_REQUEST) {
      throw new Error('expected pairing request message');
    }

    const { publicKey: peerPub, nonce: reqNonce } = await verifyPairingRequest(
      reqMsg,
      cfg.masterKey,
    );

    // 2. Check no key or label conflict in trust store
    try {
      await this.checkConflict(reqMsg.device_id, reqMsg.device_name, peerPub);
    } catch (err) {
      await this.sendRejection(transport);
      throw err;
    }

    // 3. Create and send PairingResponse
    const resp = await createPairingResponse(
      id,
      cfg.deviceName,
      cfg.capabilities,
      cfg.masterKey,
      reqNonce,
    );
    await transport.sendMessage(resp);

    // 4. Derive pairwise credential
    const respNonce = hexToBytes(resp.nonce);
    const { kPair, credRef } = await derivePairCredential(
      cfg.masterKey,
      reqNonce,
      respNonce,
      peerPub,
      id.publicKey,
    );

    // 5. Receive peer's PairingConfirm, then send ours
    const rawPeerConf = await transport.receiveMessage();
    const peerConfMsg = parsePairingMessage(rawPeerConf);
    if (!peerConfMsg || peerConfMsg.type !== MSG_PAIRING_CONFIRM) {
      throw new Error('expected pairing confirm message');
    }
    await verifyPairingConfirm(peerConfMsg, kPair, id.deviceId);

    const confirm = await createPairingConfirm(kPair, reqMsg.device_id, true);
    await transport.sendMessage(confirm);

    // 6. Record peer in local trust store & secret store (with rollback)
    const now = new Date().toISOString();
    const policy: TrustPolicy = {
      autoAccept: !!cfg.autoAccept,
      ...(cfg.autoAccept && cfg.destDir ? { autoAcceptDestDir: cfg.destDir } : {}),
      maxFileSizeBytes: 10 * 1024 * 1024 * 1024,
    };

    const record: TrustRecord = {
      deviceId: reqMsg.device_id,
      publicKey: bytesToHex(peerPub),
      localLabel: reqMsg.device_name,
      pairCredentialRef: credRef,
      relationship: RELATIONSHIP_CONTACT,
      capabilities: reqMsg.capabilities || [],
      firstSeenAt: now,
      lastSeenAt: now,
      revoked: false,
      policy,
    };

    if (this.secretStore) {
      await this.secretStore.setPairSecret(record.deviceId, kPair);
    }

    try {
      await this.trustStore.addOrUpdateDevice(record);
    } catch (err) {
      if (this.secretStore) {
        try {
          await this.secretStore.deletePairSecret(record.deviceId);
        } catch {
          // ignore rollback failure
        }
      }
      throw new Error(`save trusted device: ${(err as Error).message}`);
    }

    return {
      peerRecord: record,
      kPair,
      credRef,
    };
  }

  private async checkConflict(
    deviceId: string,
    deviceName: string,
    pubKey: Uint8Array,
  ): Promise<void> {
    // ADR 0010 §4.5: If the peer has an active tombstone record, reject pairing fail-closed
    if (this.tombstoneStore && (await this.tombstoneStore.hasTombstone(deviceId))) {
      throw new Error(ERR_TRUSTED_PEER_REVOKED);
    }

    const existing = await this.trustStore.getDevice(deviceId);
    if (existing) {
      if (existing.revoked) {
        throw new Error(ERR_TRUSTED_PEER_REVOKED);
      }
      const existingPub = hexToBytes(existing.publicKey);
      if (!constantTimeEqual(existingPub, pubKey)) {
        throw new Error(ERR_KEY_CONFLICT);
      }
    }

    // Check for active trusted devices with the same label but different device ID
    const list = await this.trustStore.listDevices();
    for (const rec of list) {
      if (
        rec.deviceId !== deviceId &&
        rec.localLabel.trim().toLowerCase() === deviceName.trim().toLowerCase() &&
        !rec.revoked
      ) {
        throw new Error(ERR_LABEL_CONFLICT);
      }
    }
  }

  private async sendRejection(transport: PairingTransport): Promise<void> {
    const rej: PairingConfirm = {
      type: MSG_PAIRING_CONFIRM,
      status: 'rejected',
    };
    try {
      await transport.sendMessage(rej);
    } catch {
      // ignore errors when sending rejection frame
    }
  }
}

File: packages/engine/trust/pairing_coordinator.go
// Package trust manages local device cryptographic identity and paired device trust records.
package trust

import (
	"bytes"
	"context"
	"crypto/ed25519"
	"encoding/hex"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/sendbeam/wire"
)

var (
	// ErrKeyConflict indicates that a pairing peer presents a conflicting public key for an existing device ID.
	ErrKeyConflict = errors.New("pairing key conflict: device ID matches an existing record with a different public key")

	// ErrLabelConflict indicates that a different active device already uses the requested local display name.
	ErrLabelConflict = errors.New("pairing conflict: a different trusted device already uses this name")
)

// PairingTransport abstracts sending and receiving pairing message payloads over an authenticated channel.
type PairingTransport interface {
	SendMessage(ctx context.Context, data []byte) error
	ReceiveMessage(ctx context.Context) ([]byte, error)
}

// PairingSessionConfig specifies options for a pairing ceremony.
type PairingSessionConfig struct {
	DeviceName   string
	Capabilities []string
	MasterKey    []byte
	AutoAccept   bool
	DestDir      string
}

// PairingResult contains the established trust record and secret pairwise credential.
type PairingResult struct {
	PeerRecord *wire.TrustRecord
	KPair      []byte
	CredRef    string
}

// PairingCoordinator drives the multi-step pairing ceremony on both initiator and responder sides.
type PairingCoordinator struct {
	idMgr     *IdentityManager
	store     Store
	credStore CredentialStore
}

// NewPairingCoordinator creates a new PairingCoordinator.
func NewPairingCoordinator(idMgr *IdentityManager, store Store) *PairingCoordinator {
	return &PairingCoordinator{
		idMgr: idMgr,
		store: store,
	}
}

// NewPairingCoordinatorWithCredentials creates a PairingCoordinator that atomically manages both device trust and credentials.
func NewPairingCoordinatorWithCredentials(idMgr *IdentityManager, store Store, credStore CredentialStore) *PairingCoordinator {
	return &PairingCoordinator{
		idMgr:     idMgr,
		store:     store,
		credStore: credStore,
	}
}

// InitiatePairing executes the initiator role of the pairing ceremony.
func (p *PairingCoordinator) InitiatePairing(ctx context.Context, transport PairingTransport, cfg PairingSessionConfig) (*PairingResult, error) {
	if transport == nil {
		return nil, errors.New("pairing transport required")
	}
	if len(cfg.MasterKey) == 0 {
		return nil, errors.New("master key required")
	}

	id, err := p.idMgr.GetOrCreateIdentity()
	if err != nil {
		return nil, fmt.Errorf("get local identity: %w", err)
	}

	// 1. Create and send PairingRequest
	req, err := wire.NewPairingRequest(id, cfg.DeviceName, cfg.Capabilities, cfg.MasterKey, nil)
	if err != nil {
		return nil, fmt.Errorf("create pairing request: %w", err)
	}

	reqData, err := wire.EncodePairingMessage(req)
	if err != nil {
		return nil, fmt.Errorf("encode pairing request: %w", err)
	}

	if err := transport.SendMessage(ctx, reqData); err != nil {
		return nil, fmt.Errorf("send pairing request: %w", err)
	}

	// 2. Receive and verify PairingResponse
	respData, err := transport.ReceiveMessage(ctx)
	if err != nil {
		return nil, fmt.Errorf("receive pairing response: %w", err)
	}

	respMsg, err := wire.DecodePairingMessage(respData)
	if err != nil {
		return nil, fmt.Errorf("decode pairing response: %w", err)
	}

	resp, ok := respMsg.(*wire.PairingResponse)
	if !ok {
		return nil, errors.New("expected pairing response message")
	}

	reqNonce, _ := hex.DecodeString(req.Nonce)
	peerPub, respNonce, err := wire.VerifyPairingResponse(resp, reqNonce, cfg.MasterKey)
	if err != nil {
		return nil, fmt.Errorf("verify pairing response: %w", err)
	}

	// 3. Derive pairwise credential
	kPair, credRef, err := wire.DerivePairCredential(cfg.MasterKey, reqNonce, respNonce, id.PublicKey, peerPub)
	if err != nil {
		return nil, fmt.Errorf("derive pair credential: %w", err)
	}

	// 4. Verify no key or label conflict in trust store
	if err := p.checkConflict(ctx, resp.DeviceID, resp.DeviceName, peerPub); err != nil {
		_ = p.sendRejection(ctx, transport)
		return nil, err
	}

	// 5. Send and receive PairingConfirm
	confirm := wire.NewPairingConfirm(kPair, resp.DeviceID, true)
	confData, err := wire.EncodePairingMessage(confirm)
	if err != nil {
		return nil, fmt.Errorf("encode pairing confirm: %w", err)
	}

	if err := transport.SendMessage(ctx, confData); err != nil {
		return nil, fmt.Errorf("send pairing confirm: %w", err)
	}

	peerConfData, err := transport.ReceiveMessage(ctx)
	if err != nil {
		return nil, fmt.Errorf("receive peer pairing confirm: %w", err)
	}

	peerConfMsg, err := wire.DecodePairingMessage(peerConfData)
	if err != nil {
		return nil, fmt.Errorf("decode peer pairing confirm: %w", err)
	}

	peerConf, ok := peerConfMsg.(*wire.PairingConfirm)
	if !ok {
		return nil, errors.New("expected pairing confirm message")
	}

	if err := wire.VerifyPairingConfirm(peerConf, kPair, id.DeviceID); err != nil {
		return nil, fmt.Errorf("verify peer pairing confirm: %w", err)
	}

	// 6. Record peer in local trust store
	now := time.Now().UTC()
	policy := wire.DefaultTrustPolicy()
	if cfg.AutoAccept && cfg.DestDir != "" {
		policy.AutoAccept = true
		policy.AutoAcceptDestDir = cfg.DestDir
	}

	record := &wire.TrustRecord{
		DeviceID:          resp.DeviceID,
		PublicKey:         hex.EncodeToString(peerPub),
		LocalLabel:        resp.DeviceName,
		PairCredentialRef: credRef,
		Relationship:      wire.RelationshipContact,
		Capabilities:      resp.Capabilities,
		FirstSeenAt:       now,
		LastSeenAt:        now,
		Revoked:           false,
		RevokedAt:         nil,
		Policy:            policy,
	}

	if p.credStore != nil {
		if err := p.credStore.SetPairSecret(ctx, record.DeviceID, credRef, kPair); err != nil {
			return nil, fmt.Errorf("persist pair credential: %w", err)
		}
	}

	if err := p.store.AddOrUpdateDevice(ctx, record); err != nil {
		if p.credStore != nil {
			_ = p.credStore.DeletePairSecret(ctx, record.DeviceID)
		}
		return nil, fmt.Errorf("save trusted device: %w", err)
	}

	return &PairingResult{
		PeerRecord: record,
		KPair:      kPair,
		CredRef:    credRef,
	}, nil
}

// AcceptPairing executes the responder role of the pairing ceremony.
func (p *PairingCoordinator) AcceptPairing(ctx context.Context, transport PairingTransport, cfg PairingSessionConfig) (*PairingResult, error) {
	if transport == nil {
		return nil, errors.New("pairing transport required")
	}
	if len(cfg.MasterKey) == 0 {
		return nil, errors.New("master key required")
	}

	id, err := p.idMgr.GetOrCreateIdentity()
	if err != nil {
		return nil, fmt.Errorf("get local identity: %w", err)
	}

	// 1. Receive and verify PairingRequest
	reqData, err := transport.ReceiveMessage(ctx)
	if err != nil {
		return nil, fmt.Errorf("receive pairing request: %w", err)
	}

	reqMsg, err := wire.DecodePairingMessage(reqData)
	if err != nil {
		return nil, fmt.Errorf("decode pairing request: %w", err)
	}

	req, ok := reqMsg.(*wire.PairingRequest)
	if !ok {
		return nil, errors.New("expected pairing request message")
	}

	peerPub, reqNonce, err := wire.VerifyPairingRequest(req, cfg.MasterKey)
	if err != nil {
		return nil, fmt.Errorf("verify pairing request: %w", err)
	}

	// 2. Check no key or label conflict in trust store
	if err := p.checkConflict(ctx, req.DeviceID, req.DeviceName, peerPub); err != nil {
		_ = p.sendRejection(ctx, transport)
		return nil, err
	}

	// 3. Create and send PairingResponse
	resp, err := wire.NewPairingResponse(id, cfg.DeviceName, cfg.Capabilities, cfg.MasterKey, reqNonce, nil)
	if err != nil {
		return nil, fmt.Errorf("create pairing response: %w", err)
	}

	respData, err := wire.EncodePairingMessage(resp)
	if err != nil {
		return nil, fmt.Errorf("encode pairing response: %w", err)
	}

	if err := transport.SendMessage(ctx, respData); err != nil {
		return nil, fmt.Errorf("send pairing response: %w", err)
	}

	// 4. Derive pairwise credential
	respNonce, _ := hex.DecodeString(resp.Nonce)
	kPair, credRef, err := wire.DerivePairCredential(cfg.MasterKey, reqNonce, respNonce, peerPub, id.PublicKey)
	if err != nil {
		return nil, fmt.Errorf("derive pair credential: %w", err)
	}

	// 5. Receive peer's PairingConfirm, then send ours
	peerConfData, err := transport.ReceiveMessage(ctx)
	if err != nil {
		return nil, fmt.Errorf("receive peer pairing confirm: %w", err)
	}

	peerConfMsg, err := wire.DecodePairingMessage(peerConfData)
	if err != nil {
		return nil, fmt.Errorf("decode peer pairing confirm: %w", err)
	}

	peerConf, ok := peerConfMsg.(*wire.PairingConfirm)
	if !ok {
		return nil, errors.New("expected pairing confirm message")
	}

	if err := wire.VerifyPairingConfirm(peerConf, kPair, id.DeviceID); err != nil {
		return nil, fmt.Errorf("verify peer pairing confirm: %w", err)
	}

	confirm := wire.NewPairingConfirm(kPair, req.DeviceID, true)
	confData, err := wire.EncodePairingMessage(confirm)
	if err != nil {
		return nil, fmt.Errorf("encode pairing confirm: %w", err)
	}

	if err := transport.SendMessage(ctx, confData); err != nil {
		return nil, fmt.Errorf("send pairing confirm: %w", err)
	}

	// 6. Record peer in local trust store
	now := time.Now().UTC()
	policy := wire.DefaultTrustPolicy()
	if cfg.AutoAccept && cfg.DestDir != "" {
		policy.AutoAccept = true
		policy.AutoAcceptDestDir = cfg.DestDir
	}

	record := &wire.TrustRecord{
		DeviceID:          req.DeviceID,
		PublicKey:         hex.EncodeToString(peerPub),
		LocalLabel:        req.DeviceName,
		PairCredentialRef: credRef,
		Relationship:      wire.RelationshipContact,
		Capabilities:      req.Capabilities,
		FirstSeenAt:       now,
		LastSeenAt:        now,
		Revoked:           false,
		RevokedAt:         nil,
		Policy:            policy,
	}

	if p.credStore != nil {
		if err := p.credStore.SetPairSecret(ctx, record.DeviceID, credRef, kPair); err != nil {
			return nil, fmt.Errorf("persist pair credential: %w", err)
		}
	}

	if err := p.store.AddOrUpdateDevice(ctx, record); err != nil {
		if p.credStore != nil {
			_ = p.credStore.DeletePairSecret(ctx, record.DeviceID)
		}
		return nil, fmt.Errorf("save trusted device: %w", err)
	}

	return &PairingResult{
		PeerRecord: record,
		KPair:      kPair,
		CredRef:    credRef,
	}, nil
}

func (p *PairingCoordinator) checkConflict(ctx context.Context, deviceID, deviceName string, pubKey ed25519.PublicKey) error {
	existing, err := p.store.GetDevice(ctx, deviceID)
	if err == nil && existing != nil {
		existingPub, err := hex.DecodeString(existing.PublicKey)
		if err == nil && !bytes.Equal(existingPub, pubKey) {
			return ErrKeyConflict
		}
	} else if err != nil && !errors.Is(err, ErrDeviceNotFound) {
		return err
	}

	// Check for active trusted devices with the same label but different device ID
	list, err := p.store.ListDevices(ctx)
	if err != nil {
		return err
	}
	for _, rec := range list {
		if rec.DeviceID != deviceID && strings.EqualFold(rec.LocalLabel, deviceName) && !rec.Revoked {
			return ErrLabelConflict
		}
	}

	return nil
}

func (p *PairingCoordinator) sendRejection(ctx context.Context, transport PairingTransport) error {
	rej := &wire.PairingConfirm{
		Type:   wire.MsgPairingConfirm,
		Status: "rejected",
	}
	data, _ := wire.EncodePairingMessage(rej)
	return transport.SendMessage(ctx, data)
}
